### PR TITLE
Focused Launch: Add DomainPicker to DomainDetail view

### PR DESCRIPTION
**Pull request by tjcafferkey**
#### Changes proposed in this Pull Request

* Add the `<DomainPicker />` component from `@automattic/domain-picker` to the `<DomainStep />` in Focused Launch
* Added `jest.config.js` to package for component tests with `@testing-library/react`
* Includes basic component tests for `<DomainStep />`
* Moved flow id to `constants.ts` into a constant named `FOCUSED_LAUNCH_FLOW_ID`

#### Prepare the sandboxed environment:
* Checkout this branch
* Make sure you have an active connection to your sandbox, and that your sandbox is clean and up to date
* In three separate terminal windows, from the root of the project, run:
  - `yarn start`
  - `cd apps/editing-toolkit && yarn dev --sync`
  - `cd apps/wpcom-block-editor && yarn dev --sync`
* Sandbox an _unlaunched_ WordPress site created via the `/start` site creation flow by adding it to your host file

#### Test for changes
* Visit `/page/SANDBOXED_SITE_ID_CREATED_WITH_START.wordpress.com/home?flags=create/focused-launch-flow`
* Click on "Launch" button — Focused Launch modal should appear
* Click on the "View all domains" link — A new placeholder view for the domain details is shown
* You should be presented with the domain picker view.
  - Domain rows should have a "Select" button on them.
  - You should have a row separating your free static sub-domain from the paid domains.
  - Changing the input value should return new premium domains.
  - Clicking "View more results" should reveal a longer list of domains.
  - Clicking "Select" on a chosen domain should change the state of that domain to "Selected".
  - Clicking "Go back" should return you to the initial screen where your selected domain is highlighted in the appropriate summary section.

#### Unit tests
* From the project root, open terminal and run `npm run test-packages`

Screenshot 2020-11-11 at 14 50 43

Fixes https://github.com/Automattic/wp-calypso/issues/46866

### CODE REVIEW

File: packages/launch/src/focused-launch/summary/index.tsx
Comment (razvanpapadopol) [inline packages/launch/src/focused-launch/summary/index.tsx:172]: Changing this to something like `summary_domain_step` would be more aligned with https://github.com/Automattic/wp-calypso/pull/47318/files#diff-a88807ae898bfacf7689a16c723644ee8d5ddf40677fd0ce119761a404d0b897R26

Comment (tjcafferkey) [reply packages/launch/src/focused-launch/summary/index.tsx]: Thanks @razvanpapadopol, done.

File: packages/launch/src/focused-launch/domain-details/test/index.tsx
Comment (razvanpapadopol) [inline packages/launch/src/focused-launch/domain-details/test/index.tsx:44]: It's really nice to see some tests ❤️ 
Of course, we can add more with time (eg: "can set domain", "can unset domain", "can fire analytics", etc) but having a starting point is more than welcome!

Comment (tjcafferkey) [reply packages/launch/src/focused-launch/domain-details/test/index.tsx:44]: Yeah, I noticed there was no tests in this package so this was more of an opportunity to implement the `jest.config.js` and the `@testing-library/react` package to enable us all to start writing them. Obviously they're rather simple tests but better than nothing right now I think 😅

File: packages/launch/src/focused-launch/domain-details/index.tsx
Comment (ciampo) [inline packages/launch/src/focused-launch/domain-details/index.tsx:94]: More of a nitpick, but can we keep calling this view "domain details" or find a name that is different from "step"? Just so that we can all agree on a naming convention.

So far, what I went with is the following:
- The focused launch modal has a router. Each route is associated to a _view_, e.g. Summary, DomainDetails, PlanDetails
- Within the Summary View, there are _steps_: the site title step, the domain step, and the plan step

If this is confusing, we can discuss and change the code / PR descriptions accordingly :)

File: packages/launch/src/focused-launch/domain-details/test/index.tsx
Comment (ciampo) [reply packages/launch/src/focused-launch/domain-details/test/index.tsx:44]: Please add more specs to #46943 , so that we can keep track of all the tests we want to write

File: packages/launch/src/focused-launch/domain-details/index.tsx
Comment (tjcafferkey) [reply packages/launch/src/focused-launch/domain-details/index.tsx:94]: This is a good point. I'll definitely make that change.

File: packages/launch/src/focused-launch/domain-details/index.tsx
Comment (ciampo) [inline packages/launch/src/focused-launch/domain-details/index.tsx:39]: We may want to "abstract" the `select`/`dispatch` calls to the `LAUNCH_STORE` in a hook, similarly to what we do with `useSite()`. This would allow us to refactor some code in the Summary view too.

Although I know we also have issue #47233 — @razvanpapadopol , do you think the refactor to a `useLaunch()` hook should happen in this PR, or when working on #47233?

File: packages/launch/src/focused-launch/domain-details/test/index.tsx
Comment (ciampo) [inline packages/launch/src/focused-launch/domain-details/test/index.tsx:35]: Should we test for the back button to point to the correct view?

**Review by ciampo — DISMISSED** [review]
> - Clicking "Select" on a chosen domain should change the state of that domain to "Selected".
> - Clicking "Go back" should return you to the initial screen where your selected domain is highlighted in the appropriate summary section.

We need to sync the domain step in the summary view, with the currently selected domain:
- If I go into the domain details view,
- I click "show more" and I select a domain from the ones that were just revealed
- I click "Go back"
- The domain step doesn't show the domain that I just selected (instead it keeps showing the same 3 suggestions as it did before)

A similar problem happens if I go back to the Domain Details view: since the domain that I selected is under the "more domains" fold, when I first enter the view I don't see anything selected, and I'm not reminded of what is my currently selected domain

cc @razvanpapadopol

File: packages/launch/src/focused-launch/domain-details/test/index.tsx
Comment (tjcafferkey) [reply packages/launch/src/focused-launch/domain-details/test/index.tsx:35]: This test case is reflected in https://github.com/Automattic/wp-calypso/issues/46943

Comment (ciampo) [reply packages/launch/src/focused-launch/domain-details/test/index.tsx:35]: Cool, thank you! Not a blocker for this PR

File: packages/launch/package.json
Comment (ciampo) [inline packages/launch/package.json:33]: We should probably also link to the `calypso-analytics` package in `packages/launch/tsconfig.json`, under the `references` key

Comment (tjcafferkey) [reply packages/launch/package.json:33]: Done

**Comment by tjcafferkey** [comment]
> A similar problem happens if I go back to the Domain Details view: since the domain that I selected is under the "more domains" fold, when I first enter the view I don't see anything selected, and I'm not reminded of what is my currently selected domain

@ciampo Yes I can see the issue here. This also happens on Gutenboarding when returning to the domains view if you have selected a domain within the "View more results" section of the domain picker.

See [video](https://www.loom.com/share/ec8d6d12b4e8496391659c07330faff5) for example

File: packages/launch/src/focused-launch/domain-details/index.tsx
Comment (tjcafferkey) [reply packages/launch/src/focused-launch/domain-details/index.tsx:39]: >We may want to "abstract" the select/dispatch calls to the LAUNCH_STORE in a hook, similarly to what we do with useSite(). This would allow us to refactor some code in the Summary view too.

As per our brief Slack discussion, we are in agreement that we should do this within this PR 👍🏻

**Comment by ciampo** [comment]
> > A similar problem happens if I go back to the Domain Details view: since the domain that I selected is under the "more domains" fold, when I first enter the view I don't see anything selected, and I'm not reminded of what is my currently selected domain
> 
> @ciampo Yes I can see the issue here. This also happens on Gutenboarding when returning to the domains view if you have selected a domain within the "_View more results_" section of the domain picker.
> 
> See [video](https://www.loom.com/share/ec8d6d12b4e8496391659c07330faff5) for example

In order to unblock this PR, we could track these 2 issues separately:

- Always show the currently selected domain in the domain details view
- Always show the currently selected domain in the domain step within the summary view

If we agree, I'll proceed and create 2 separate issues

File: packages/launch/src/focused-launch/domain-details/index.tsx
Comment (razvanpapadopol) [reply packages/launch/src/focused-launch/domain-details/index.tsx]: > We may want to "abstract" the select/dispatch calls to the LAUNCH_STORE in a hook, similarly to what we do with useSite(). This would allow us to refactor some code in the Summary view too.
>
> Although I know we also have issue #47233 — @razvanpapadopol , do you think the refactor to a useLaunch() hook should happen in this PR, or when working on #47233?

I strongly encourage abstracting bundles of multiple selectors and/or additional logic to a hook that is reusable between flows (focused and step-by-step launch in this case) and the perfect example would be `use-domain-search`. 

But at the same time, there is no real benefit from extracting just a select/dispatch call if there is no extra processing that we could include in the abstraction.

**Comment by razvanpapadopol** [comment]
> In order to unblock this PR, we could track these 2 issues separately:

I see one "Domain Picker" issue where we'll need to check internally within the domain-picker package if the `currentDomain` (passed as prop) is not inside the [domainSuggestions array](https://github.com/Automattic/wp-calypso/blob/master/packages/domain-picker/src/domain-picker/index.tsx#L140). In this case we should replace the first/last suggestion with `currentDomain`. But we should keep in mind that `currentDomain` is currently a string and not a `DomainSuggestion` object.

- Focused launch: https://github.com/Automattic/wp-calypso/pull/47318#pullrequestreview-528950382

> Always show the currently selected domain in the domain details view
Always show the currently selected domain in the domain step within the summary view

☝️ Let's also make sure we're still showing the free sub-domain on the summary view.

- New onboarding (previously called Gutenboarding): https://github.com/Automattic/wp-calypso/pull/47318#issuecomment-726007390

File: packages/launch/src/focused-launch/domain-details/index.tsx
Comment (razvanpapadopol) [inline packages/launch/src/focused-launch/domain-details/index.tsx:69]: We want to display directly the expanded view in DomainDetails.

```suggestion
					itemType={ ITEM_TYPE_BUTTON }
					quantity={ 10 }
```


This should also fix what @ciampo was reporting previously:
> A similar problem happens if I go back to the Domain Details view: since the domain that I selected is under the "more domains" fold, when I first enter the view I don't see anything selected, and I'm not reminded of what is my currently selected domain

**Review by razvanpapadopol — CHANGES_REQUESTED** [review]

**Comment by ciampo** [comment]
Also, just a heads up — we should rebase this PR (or merge master into it) before approving and merging, as the changes to `useDomainSearch()` hook introduced in #47528  may break the code in this PR

**Comment by tjcafferkey** [comment]
>Also, just a heads up — we should rebase this PR (or merge master into it) before approving and merging, as the changes to useDomainSearch() hook introduced in #47528 may break the code in this PR

Waiting for this PR (#47528) to get merged so I can rebase before marking this as "Ready For Review" again

**Review by alshakero — APPROVED** [review]
Tested and reviewed. The only unmet criterion is this

> Clicking "View more results" should reveal a longer list of domains.

I didn't find the <kbd>Show more</kbd> button.

**Comment by razvanpapadopol** [comment]
> Tested and reviewed. The only unmet criterion is this
> 
> > Clicking "View more results" should reveal a longer list of domains.
> 
> I didn't find the Show more button.

Checked Figma now and by default, we should display the Domain Picker not expanded. 

I've [previously requested it to be expanded](https://github.com/Automattic/wp-calypso/pull/47318#discussion_r522130477) so the selection to be visible at all times. But now this will be fixed as the 3rd acceptance crietria in https://github.com/Automattic/wp-calypso/issues/47512

**Review by razvanpapadopol — APPROVED** [review]

**Pull request merged**
